File: backtracking/131.go
func partition(s string) [][]string {
    ret := make([][]string, 0)
    if s == "" {
        return ret
    }
    dfs(&ret, 0, s, []string{})
    return ret
}

func dfs(ret *[][]string, start int, origin string, cur []string) {
    if start == len(origin) {
        *ret = append(*ret, append([]string{}, cur...))
        return
    }
    for i := start+1; i <= len(origin); i++ {
        // 对于每一个正确的prefix, dfs剩余的string
        if isPalindrome(origin[start:i]) {
            temp := append(append([]string{}, cur...), origin[start:i])
            dfs(ret, i, origin, temp)
        }
    }
}

func isPalindrome(s string) bool {
    for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
        if s[i] != s[j] {
            return false
        }
    }
    return true 
}

File: backtracking/131.cc
class Solution {
public:
    vector<vector<string>> partition(string s) {
        std::vector<std::vector<std::string>> ret;
        std::vector<std::string> cur;
        recur(ret, s, 0, cur);
        return ret;
    }
private:
    void recur(std::vector<std::vector<std::string>>& ret, string s, int index, std::vector<std::string>& cur) {
        if (index == s.size()) {
            // ret.emplace_back(std::move(cur));
            ret.emplace_back(cur);
            return;
        }
        for (int i = index+1; i <= s.size(); ++i) {
            std::string temp = s.substr(index, i-index);
            if (isPalin(temp)) {
                cur.emplace_back(std::move(temp));
                recur(ret, s, i, cur);
                cur.pop_back();
            }
        }
    }
    bool isPalin(std::string& s) {
        for (int i = 0, j = s.size()-1; i < j; ++i, --j) {
            if (s[i] != s[j])
                return false;
        }
        return true;
    }
};
